relays/messages/src/relay_strategy/altruistic_strategy.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Altruistic relay strategy
+
+use async_trait::async_trait;
+
+use bp_messages::{MessageNonce, Weight};
+use bp_runtime::messages::DispatchFeePayment;
+
+use crate::{
+	message_lane::MessageLane,
+	message_lane_loop::{
+		SourceClient as MessageLaneSourceClient, TargetClient as MessageLaneTargetClient,
+	},
+	message_race_loop::NoncesRange,
+	relay_strategy::{RelayReference, RelayStrategy},
+};
+
+/// The relayer doesn't care about rewards.
+#[derive(Clone)]
+pub struct AltruisticStrategy;
+
+#[async_trait]
+impl RelayStrategy for AltruisticStrategy {
+	async fn decide<
+		P: MessageLane,
+		SourceClient: MessageLaneSourceClient<P>,
+		TargetClient: MessageLaneTargetClient<P>,
+	>(
+		reference: RelayReference<P, SourceClient, TargetClient>,
+	) -> Option<MessageNonce> {
+		let mut hard_selected_count = 0;
+		let mut soft_selected_count = 0;
+
+		let mut selected_weight: Weight = 0;
+		let mut selected_unpaid_weight: Weight = 0;
+		let mut selected_prepaid_nonces = 0;
+		let mut selected_size: u32 = 0;
+		let mut selected_count: MessageNonce = 0;
+
+		let hard_selected_begin_nonce =
+			reference.nonces_queue[reference.nonces_queue_range.start].1.begin();
+
+		let all_ready_nonces = reference
+			.nonces_queue
+			.range(reference.nonces_queue_range.clone())
+			.flat_map(|(_, ready_nonces)| ready_nonces.iter())
+			.enumerate();
+		for (index, (_nonce, details)) in all_ready_nonces {
+			// Since we (hopefully) have some reserves in `max_messages_weight_in_single_batch`
+			// and `max_messages_size_in_single_batch`, we may still try to submit transaction
+			// with single message if message overflows these limits. The worst case would be if
+			// transaction will be rejected by the target runtime, but at least we have tried.
+
+			// limit messages in the batch by weight
+			let new_selected_weight = match selected_weight.checked_add(details.dispatch_weight) {
```

Comment:
There are two types of checks in this loop - `hard` and `soft`. You can't violate hard checks in any strategy - i.e. if target chain weight has max transaction size set to 1MB, it means that you can't deliver 2MB of messages in a single tx. That's the 'hard' limit. The 'soft' limit is defined by the strategy. So I'd expect `*_strategy.rs` files to only contain soft checks. Hard checks must be done at the upper level and stay the same for all strategies.

That said, I expect `altruistic_strategy.rs` to only have something like `return true` or `return <something-that-says-Im-ready-to-deliver-all-nonces>` in its only method. And the `rational_strategy.rs` should only have checks related to rewards/losses. Is it possible to fix that, please?

Reply:
I changed it. can be reviewed.

Reply:
Oh, sorry, This is still wrong, please wait. let me fix it

Reply:
Done

Reply:
I don't know if there is a better solution, there is currently a big struct 
https://github.com/darwinia-network/parity-bridges-common/blob/845ac793b346aa594e0afcdff1d8ca62ca46d90c/relays/messages/src/relay_strategy/enforcement_strategy.rs?#L64-L89